### ISSUE TITLE
Welcome page and interactive playground code should lazy load when the feature is disabled

### ISSUE THREAD

**Issue by Tyriar** [microsoft/vscode#32980]
Quite a lot of welcome page code appears to be getting loaded even when the feature is disabled. We should avoid loading the majority of the code on start up but defer to when/if it's used.

<img width="506" alt="screen shot 2017-08-22 at 1 26 22 pm" src="https://user-images.githubusercontent.com/2193314/29585893-9acaf8da-873d-11e7-93ca-157785f52027.png">

Dynamic imports could help: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html

/cc @jrieken

**Comment by chrmarti** [comment]
@jrieken Aren't we loading everything as a single minified filed?

**Comment by Tyriar** [comment]
@jrieken I'm also curious on your advice around the best way to do this. I've been throwing around the idea of applying this to the entire terminal part (excluding the contribution file), not just xterm.js. May need improvements in the build system to build minified/concated sub-components though.

That's how I originally thought everything worked; components are loaded only when they're needed. The truth appears to be that the code is still loaded/parsed though which will continue to slow start up/increase memory consumption for components that may never be used (such as the terminal and welcome/playground).

**Comment by jrieken** [comment]
Yeah, it is tricky. Code loading has the biggest impact on startup performance and perf is one of the most popular feedback items. We have some tricks in place, like bundling, minification. On my machine, code loading takes ~1,2seconds, when using v8 cached data half of that. Cached data is used on first startup and for new windows, tho doesn't work on "reload". 

Considering the huge of code we are loading, this actually isn't bad but looking at the rate of which we write new code we should be prepared. I have made a push to lazy-load npm-modules, they take around 200ms of the code loading time and they are *easy* to extract - there is now TypeScript support and there is no bundling configuration required.

After lazy loading npm-modules the next logical step would be to lazy load amd-modules as well. It does require work because you need to statically describe what subtree to bundle in separate files. We have done that in the early times, e.g. each language, each viewlet was its own bundle. The challenge is to keep your dependencies minimal. The reason why we have `xyz.contribution.ts`-files is that we wanted those to have few dependencies and to contain interfaces only such that other parts of the system only dependent on a small subset, think of the explorer adding the 'Find In This Folder' action.  Tho we were never really strict and laziness made those files grow and dependencies getting larger and larger. In the end only the UI pieces were bundled separately, continuing the sample: only the search viewlet not its service, model, commands, etc. In short the savings didn't pay off given the costs of maintaining the bundle configuration files. 

Long story short is that it isn't easy and only paying off if done properly. I wouldn't do an at hoc change because this requires some more thinking...

**Comment by chrmarti** [comment]
Sounds good. Let us know when there is an #issue for the general discussion.
